### ISSUE TITLE
[Critical] V4  - Query by content relationship field not working

### ISSUE THREAD

**Issue by kb1995** [prismicio/gatsby#414]
### Versions

- gatsby-source-prismic: v4.0.1
- gatsby-plugin-prismic-previews: ^4.0.1
- node: v14.17.1

### Reproduction

Follow the tutorial https://prismic.io/docs/technologies/query-by-content-relationship-link-field-graphql

I have a custom type `Academy Post` and I have a content relationship "author". I want to create a page /author/:uid where I can extract all academy posts with this author. However, whenever I try to do that, I get the following error in the graphql

```
{
  "errors": [
    {
      "message": "Unknown argument \"where\" on field \"Query.allPrismicAcademyPost\".",
      "locations": [
        {
          "line": 2,
          "column": 25
        }
      ],
      "extensions": {
        "stack": [
          "GraphQLError: Unknown argument \"where\" on field \"Query.allPrismicAcademyPost\".",
          "    at Object.Argument (/Users/kris/Documents/code/freelance/sharetribe/nemesis/node_modules/graphql/validation/rules/KnownArgumentNamesRule.js:46:29)",
          "    at Object.enter (/Users/kris/Documents/code/freelance/sharetribe/nemesis/node_modules/graphql/language/visitor.js:323:29)",
          "    at Object.enter (/Users/kris/Documents/code/freelance/sharetribe/nemesis/node_modules/graphql/utilities/TypeInfo.js:370:25)",
          "    at visit (/Users/kris/Documents/code/freelance/sharetribe/nemesis/node_modules/graphql/language/visitor.js:243:26)",
          "    at validate (/Users/kris/Documents/code/freelance/sharetribe/nemesis/node_modules/graphql/validation/validate.js:69:24)",
          "    at graphqlMiddleware (/Users/kris/Documents/code/freelance/sharetribe/nemesis/node_modules/express-graphql/index.js:98:38)",
          "    at runMicrotasks (<anonymous>)",
          "    at processTicksAndRejections (internal/process/task_queues.js:95:5)"
        ]
      }
    }
  ]
}
```

My GraphQl query looks like this:

```
query MyQuery {
  allPrismicAcademyPost(where: {author: "ab965d0c-20bd-5e37-ad48-0be4c733c59a"}) {
    edges {
      node {
        id
      }
    }
  }
}
```

I'm stuck at this point, as I'm not sure how to resolve this. Is this a bug or a user error?

I'm using the custom types API to automatically retrieve all custom types

**Comment by angeloashmore** [comment]
Hey @kb1995, the way to filter your queries is different in Gatsby's GraphQL API compared to Prismic's GraphQL API.

Gatsby has its own way of making its data available to you. If you source content from Prismic and other APIs, such as Shopify, Gatsby will make querying data for both available using the same interface.

This means the API will be different than when using Prismic's native GraphQL API.

The query you posted can be done in Gatsby using the following query:

```graphql
query MyQuery {
  allPrismicAcademyPost(
    filter: {data: {author: {id: {eq: "ab965d0c-20bd-5e37-ad48-0be4c733c59a"}}}}
  ) {
    nodes {
      id
    }
  }
}
```

You can reference the following documentation to learn how to query in Gatsby:

- https://prismic.io/docs/technologies/query-basics-gatsby
- https://www.gatsbyjs.com/docs/how-to/querying-data/

I would recommend building your querying using Gatsby's built-in GraphiQL tool. You can learn more here: https://www.gatsbyjs.com/docs/how-to/querying-data/running-queries-with-graphiql/

**Comment by kb1995** [comment]
@angeloashmore Hey, thanks a lot for the detailed response! The query that you posted could work, I will give it a go tomorrow!

However, I'm still a bit confused why the `where` argument described in [this article](https://prismic.io/docs/technologies/query-by-content-relationship-link-field-graphql) is not working and how I could make it work with my Prismic custom type and my graphql query. Is this something specifically for the plugin's backend and not accessible to developers?

**Comment by angeloashmore** [comment]
@kb1995 Hopefully I can explain clearly, but let me know if I should expand on anything.

The reason is that they are two different APIs. Although both use GraphQL as their way of querying for data, they do not use GraphQL in the same way.

## Prismic's REST and GraphQL APIs

When using Prismic outside of a Gatsby application, content can be fetched using Prismic's REST API or GraphQL API.

<img src="https://user-images.githubusercontent.com/8601064/127241030-4027f071-0f48-440e-ab47-3d334316f606.png" width="400" />

For example, a React app or an Express server may fetch content via Prismic's REST API while an iOS app uses the GraphQL API. Those users would use the following docs:

- **REST API Docs**: https://prismic.io/docs/technologies/introduction-to-the-content-query-api
- **GraphQL API Docs**: https://prismic.io/docs/technologies/intro-to-the-graphql-api

## Gatsby's GraphQL API

When using Prismic in an Gatsby application, content is fetched using Gatsby's GraphQL API. `gatsby-source-prismic` makes this possible by first fetching all content from your Prismic repository using Prismic REST API. It adds your content into Gatsby's internal database, which then makes it available via your page queries.

<img src="https://user-images.githubusercontent.com/8601064/127241164-10a019d9-9e80-4053-9a32-9731b0b1d645.png" width="400" />

These users would use the following docs:

- **Prismic's Gatsby Docs**: https://prismic.io/docs/technologies/gatsby
- **Prismic's Gatsby Querying Docs**: https://prismic.io/docs/technologies/query-basics-gatsby

Specifically for filtering data, like what you are trying to do with `where`, you can read this page:

- **Arguments: filter, sort, limit & skip (Gatsby)**: https://prismic.io/docs/technologies/arguments-filter-sort-limit-skip-gatsby

Basically, when using Gatsby, *do not* use Prismic's GraphQL API documentation because that is not what you use with Gatsby.

Hope that helps!

**Comment by kb1995** [comment]
@angeloashmore That's a stellar explanation - makes perfect sense now. Thanks a lot!
